Decouple compaction manager from database

bhalevy opened the pull request:
Start compaction_manager as a sharded service
and pass a reference to it to the database rather
than having the database construct its own compaction_manager.

This is part of the wider scope effort to decouple compaction from replica database and table.

nyh commented:
> Start compaction_manager as a sharded service and pass a reference to it to the database rather than having the database construct its own compaction_manager.
> 
> This is part of the wider scope effort to decouple compaction from replica database and table.

Maybe this was discussed in the past but I missed this discussion and I'm curious about the motivation. You say there's a "wider scope effort to decouple compaction from replica database and table". But why? What can a compaction_manager do without a database? On the other hand, a database can live fine (though inefficiently) without compaction, so it sort of makes sense for a database to start its compaction manager after it starts, and shut the compaction manager down before the database shuts itself down.

So what was the motivation here?

xemul commented:
@nyh , there's a circular dependency there -- compaction manager needs system keyspace (to INSERT/UPDATE its stats into), system keyspace (today) needs a database and database needs compaction manager. @bhalevy , does this set helps breaking this loop?

xemul commented on compaction/compaction_manager.cc:
```diff
@@ -622,10 +622,11 @@ sstables::compaction_stopped_exception compaction_manager::task::make_compaction
-    : _compaction_sg(cfg.compaction_sched_group)
-    , _maintenance_sg(cfg.maintenance_sched_group)
-    , _compaction_controller(make_compaction_controller(_compaction_sg, cfg.static_shares, [this, available_memory = cfg.available_memory] () -> float {
+compaction_manager::compaction_manager(std::function<config()> get_cfg, abort_source& as)
```
Config getter :scream: 
I guess it's because of #7316 ?

bhalevy replied:
Yes, exactly.

xemul replied:
Awesome! The a246b6d3eb0de2bb359b32a269b7247d20ca9fbe suffers from the same, so likely the issue is going to be fixed really soon :)

avikivity replied:
Ouch

xemul commented on compaction/compaction_manager.hh:
```diff
@@ -48,9 +48,9 @@ public:
-        scheduling_group compaction_sched_group;
-        scheduling_group maintenance_sched_group;
-        size_t available_memory;
+        scheduling_group compaction_sched_group = scheduling_group{default_scheduling_group(), default_priority_class()};
```
Let's better patch `backlog_controller::scheduling_group` to have default-initialized parts? One of them is a reference, but it can be safely converted into a value, the `io_priority_class` is just a handle (with the class index) that can be copied around pretty much everywhere without any side effects.

xemul commented on replica/database.cc:
```diff
@@ -81,15 +81,7 @@ logging::logger dblog("database");
-
-inline
-flush_controller
-make_flush_controller(const db::config& cfg, backlog_controller::scheduling_group& sg, std::function<double()> fn) {
```
nit: Likely the `make_flush_controller` move deserves its own patch to reduce the churn here

bhalevy replied:
Actually I moved `make_compaction_manager_config` out of the replica namespace, and didn't touch make_flush_controller (because the logic really belongs to compaction_manger).
I can do that in a preliminary patch so to reduce thew churn here.

avikivity commented:
> @nyh , there's a circular dependency there -- compaction manager needs system keyspace (to INSERT/UPDATE its stats into), system keyspace (today) needs a database and database needs compaction manager. @bhalevy , does this set helps breaking this loop?


This can be decoupled by passing a function (or interface) to perform the stats update.

xemul commented on scylla-gdb.py:
```diff
@@ -4605,7 +4605,7 @@ def __init__(self):
     def invoke(self, arg, from_tty):
         db = find_db()
-        cm = std_unique_ptr(db['_compaction_manager']).get().dereference()
+        cm = db['_compaction_manager']
```
Is this backward-compatible? I guess it should look somehow like

```python
    cm = db['_compaction_manager']
    try:
        aux = std_unique_ptr(db['_compaction_manager']).get().dereference()
        cm = aux
    except:
        pass
```

bhalevy replied:
Right, sorry.  You're right.

xemul commented:
> This can be decoupled by passing a function (or interface) to perform the stats update.

Yes, this would work. Same (both -- the loop and the proposed scissors) is true for large data handler, btw.

raphaelsc commented on compaction/compaction_manager.hh:
```diff
@@ -482,6 +484,11 @@ public:
+namespace replica { class database_config; }
+
+compaction_manager::config make_compaction_manager_config(const db::config& cfg, replica::database_config& dbcfg);
+
```
looks like this function could be declared in replica/database.hh instead, so we don't have to forward declare those things from db and replica namespaces.

raphaelsc replied:
what if we expand the list of params in make_compaction_manager_config? we want to stop including replica/database.hh from compaction_manager.cc. if we're declaring make_compaction_manager_config() in compaction_manager.hh, so its definition should be moved to compaction_manager.cc.

raphaelsc replied:
but we can't because there's updatable static shares :-( we'll need to think about something.

bhalevy replied:
Maybe I'll just open code it in main and cql_test_env (as I did in an earlier draft that I didn't publish).
This will duplicate the code in both places but the current way it's done is awkward and really our of scope of both replica::database and compaction_manager.  The config stuff belongs to the application layer.

xemul replied:
If you open-code it into main.cc, it will be "duplicated" only between main and cql_test_env -- the same way as pretty much every other service config is (also mind #2705), so I'm for open-coding this helper

bhalevy replied:
Right.  We want to unify initialization anyhow.
Seems like you mentioned the wrong issue though... You probably meant #2795

raphaelsc commented on main.cc:
```diff
@@ -947,10 +948,14 @@ To start the scylla server proper, simply invoke as: scylla server (or just scyl
 
+            supervisor::notify("starting compaction_manager");
+            cm.start([&] { return make_compaction_manager_config(*cfg, dbcfg); }, std::ref(stop_signal.as_sharded_abort_source())).get();
+            auto stop_cm = deferred_stop(cm);
```
the reason for lambda is to retrieve configuration from each shard, right? but reading make_compaction_manager_config() it looks like we don't have to. we're using the same parameters at all shards. could you please explain the motivation?

xemul replied:
@raphaelsc , I believe it's due to #7316

bhalevy replied:
We need to in order to get the per-shared updateable value in
https://github.com/scylladb/scylla/blob/39db15d2cba81870bcc8269ff8afebb2fe2d2038/utils/config_file.hh#L155-L158

denesb replied:
We have `sharded_parameter` for this purpose exactly.

raphaelsc commented:
> > @nyh , there's a circular dependency there -- compaction manager needs system keyspace (to INSERT/UPDATE its stats into), system keyspace (today) needs a database and database needs compaction manager. @bhalevy , does this set helps breaking this loop?
> 
> This can be decoupled by passing a function (or interface) to perform the stats update.

compaction manager is now updating system table through table_state interface.

raphaelsc commented:
> > > @nyh , there's a circular dependency there -- compaction manager needs system keyspace (to INSERT/UPDATE its stats into), system keyspace (today) needs a database and database needs compaction manager. @bhalevy , does this set helps breaking this loop?
> > 
> > 
> > This can be decoupled by passing a function (or interface) to perform the stats update.
> 
> compaction manager is now updating system table through table_state interface.

once we remove replica::database usage from perform_cleanup() and perform_sstable_upgrade(), then compaction_manager.cc can stop including replica/database.hh. @bhalevy could you please include these changes in this series?

avikivity commented:
> > This can be decoupled by passing a function (or interface) to perform the stats update.
> 
> Yes, this would work. Same (both -- the loop and the proposed scissors) is true for large data handler, btw.

large data handler is already decoupled, no? it uses an interface.

It's coupled in the source because cql_large_data_handler is in the same .hh as large_data_handler, but that's a cut'n'paste away.

bhalevy commented on replica/database.cc:
```diff
@@ -2237,14 +2237,13 @@ void database::revert_initial_system_read_concurrency_boost() {
     _large_data_handler->start();
     // We need the compaction manager ready early so we can reshard.
-    _compaction_manager->enable();
+    _compaction_manager.enable();
```
I think this deserves a followup patch, not sure if it fit in the scope of this mini-series.

xemul replied:
Sure

xemul commented:
> large data handler is already decoupled, no? it uses an interface.

It uses `qctx` :nauseated_face:  and makes direct requests to `system.large_{}s` described in system keyspace .cc so I'd count it as not-yet-explicitly-patched dependency

> It's coupled in the source because cql_large_data_handler is in the same .hh as large_data_handler, but that's a cut'n'paste away.

`cql_large_data_handler`?! What is it?

scylladb-promoter commented:
CI state **`SUCCESS`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/1466/

bhalevy commented:
In v4 (68551a1405):  (v1 and v2 were sent to the scylladb-dev mailing list, v2 is the submitted PR)
- rebased
- Added:
  - backlog_controller: scheduling_group: define default member initializers
  - backlog_controller: scheduling_group: keep io_priority_class by value
- main: start compaction_manager as a sharded service
  - open coded `make_compaction_manager_config` in main and cql_test_env
  - made the change in scylla-gdb.py backward compatible as per https://github.com/scylladb/scylla/pull/11099#discussion_r926624461
- Added:
  - compaction: cleanup, upgrade: use a lw_shared_ptr for owned token ranges
  - compaction_manager: perform_cleanup, perform_sstable_upgrade: use a lw_shared_ptr for owned token ranges
    - deletes `#include "replica/database.hh"` from compaction_manager.cc

scylladb-promoter commented:
CI state **`SUCCESS`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/1476/

xemul commented on compaction/compaction_manager.hh:
```diff
@@ -275,6 +275,7 @@ private:
     timer<lowres_clock> _compaction_submission_timer = timer<lowres_clock>(compaction_submission_callback());
     static constexpr std::chrono::seconds periodic_compaction_submission_interval() { return std::chrono::seconds(3600); }
 
+    config _cfg;
```
Since the whole config now lives on compaction manager, it's worth removing all the duplicates like `_compaction_sg`, `_maintenance_sg`, `_throughput_mbs`, `_static_shares` and `_available_memory`

bhalevy replied:
makes sense

bhalevy replied:
Done in v5

bhalevy commented:
In v5 (cc82cac3b6)
- backlog_controller: keep scheduling_group by value
  - to simplify next patch that uses const getters for the config / scheduling group
- compaction_manager: keep config as member
  - eliminates duplicate members

scylladb-promoter commented:
CI state **`SUCCESS`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/1489/

raphaelsc commented on compaction/compaction_manager.hh:
```diff
@@ -362,6 +358,26 @@ public:
+        return _cfg.static_shares.get();
+    }
+
+    auto throughput_mbs() const noexcept {
```
minor: it's better to restrict automatic return type for complex types, for example:

```
template <typename Func, typename T>
auto future_invoke(Func&& func, T&& v) {
    if constexpr (std::is_same_v<T, monostate>) {
        return std::invoke(std::forward<Func>(func));
    } else {
        return std::invoke(std::forward<Func>(func), std::forward<T>(v));
    }
}
```
the return type is of great help to document a function, the user knows what's expected, so better use it whenever possible.

denesb replied:
Agree, we should avoid using auto in return and parameter types just for convenience.

bhalevy replied:
The main problem with that is that if the member type changes and the get method may implicitly convert it to the old type with no warning, while `auto` convey exactly what we want to do - i.e. return the value or a reference to it, whatever type it is.

raphaelsc replied:
> The main problem with that is that if the member type changes and the get method may implicitly convert it to the old type with no warning, while `auto` convey exactly what we want to do - i.e. return the value or a reference to it, whatever type it is.

That same problem can still happen to the users of the interface. I think auto as return type should be restricted for complex types where template is involved.

Otherwise every class out there having an integer as member could also use auto when exposing it through its public interface. The return type is very important for documenting the interface too.

If you change a interface from signed to unsigned, then you must be careful that every user of the interface won't break.

bhalevy replied:
Done in v6

raphaelsc commented on scylla-gdb.py:
```diff
@@ -4605,7 +4605,12 @@ def __init__(self):
+        cm = db['_compaction_manager']
+        try:
+            aux = std_unique_ptr(db['_compaction_manager']).get().dereference()
+            cm = aux
```
nice trick :-)

bhalevy replied:
It was @xemul's idea IIRC :)

raphaelsc commented on api/storage_service.cc:
```diff
@@ -642,10 +642,11 @@ void set_storage_service(http_context& ctx, routes& r, sharded<service::storage_
                 });
                 auto& cm = db.get_compaction_manager();
+                auto owned_ranges_ptr = compaction::make_owned_ranges_ptr(db.get_keyspace_local_ranges(keyspace));
                 // as a table can be dropped during loop below, let's find it before issuing the cleanup request.
```
we're not supposed to change topology while cleanup is running (right?), so calculating ranges once before all cleanup, and not once before each cleanup job, is not a problem.

bhalevy replied:
> we're not supposed to change topology while cleanup is running (right?)

Unfortunately, there is nothing to prevent it.
We're still relying on the admin's good will until we enforce it using transactional topology changes, and including cleanup as an integral part of the node(s)-operation.

bhalevy replied:
> calculating ranges once before all cleanup, and not once before each cleanup job, is not a problem.

I'm not sure what you're saying exactly.
The owned_ranges depend on the effective replication strategy, therefore our internal api for getting them is per keyspace.

raphaelsc replied:
some history context: previously we called the api before each sstables::compact_sstables(), today we call the api on each perform_cleanup(), and next, we'll do it once before each keyspace. This change caught my attention.

But like I said, it's a not problem as we expect no topology change while cleanup is running, and like you said, we'll later enforce it through raft.

raphaelsc reviewed (approved):
had only one minor comment. change be postponed though, shouldn't block this series.

denesb commented on backlog_controller.hh:
```diff
@@ -78,7 +78,7 @@ protected:
     void adjust();
 
-    backlog_controller(scheduling_group& sg, std::chrono::milliseconds interval,
+    backlog_controller(const scheduling_group& sg, std::chrono::milliseconds interval,
```
nit: should probably be just a value at this point

bhalevy replied:
Yeah, it should work

bhalevy replied:
done in v6

denesb commented on compaction/compaction_manager.hh:
```diff
@@ -352,7 +352,9 @@ private:
-    compaction_manager(config cfg, abort_source& as);
+    // get_cfg is called on each shard when starting a sharded<compaction_manager>
+    // we need the getter since updateable_value is not shard-safe (#7316)
+    compaction_manager(std::function<config()> get_cfg, abort_source& as);
```
Isn't this https://docs.seastar.io/master/classseastar_1_1sharded__parameter.html open-coded?

bhalevy replied:
yes.  done in v6

denesb commented on replica/database.cc:
```diff
@@ -2671,7 +2660,7 @@ future<> database::flush_system_column_families() {
     auto b = defer([this] { _stop_barrier.abort(); });
     // Interrupt on going compaction and shutdown to prevent further compaction
-    co_await _compaction_manager->drain();
+    co_await _compaction_manager.drain();
```
`database::shutdown()` doesn't call `compaction_manager.stop()` anymore. So why is `drain()` still called by database?

bhalevy replied:
drain on shutdown is still a bastard.
it's called on the storage_service that in turn calls database::drain which calls compaction_manager::drain (among other things).

I think it's possible to promote the compaction_manager.drain part to be called from the upper layer, but we still want to do it out of order (i.e. not in reverse to the services start order).  Since compaction can be stopped at any time, it shouldn't be a problem to do it first before shutting down / draining other services that depend on storage.

xemul replied:
The drain-on-shutdown is just a boring wrapper to drain the node on shutdown if it wasn't explicitly drained before. Whether or not to put compaction_manager.drain() here or "standalone" is rather decided from -- should it be called on the `nodetool drain` or not. If not, then it's possible to take some care and move it to the upper layer. If yes -- it should stay here, at least for now.

denesb commented on backlog_controller.hh:
```diff
@@ -39,7 +39,7 @@ class backlog_controller {
     struct scheduling_group {
         seastar::scheduling_group cpu = default_scheduling_group();
-        const ::io_priority_class& io = default_priority_class();
+        seastar::io_priority_class io = default_priority_class();
```
At this point the initializers are not needed because their default constructor is equivalent to the `default_*()` factory method result, no?

xemul replied:
Only the `scheduling_group` has. But `io_priority_class` [is going away](https://github.com/scylladb/seastar/issues/1069) anyway

bhalevy commented:
In v6 (435cbe642b):
- rebased + retested
- backlog_controller: keep scheduling_group by value
  - pass scheduling_group by value to constructors as well
- compaction_manager: keep config as member
  - use explicit return types for get methods
- main: start compaction_manager as a sharded service
  - simplify using sharded_parameter

scylladb-promoter commented:
CI state **`FAILURE`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/1571/

xemul reviewed (approved):
CI fails because of java package. @bhalevy , I guess it makes sense just to rebase and repush so that CI picks up fresh java submodule and reruns (and passes)

bhalevy commented:
b4a7c825c4: rebased

denesb commented:
Conflicts, please rebase again.

bhalevy commented:
14faa3b6f4: rebased and retested

scylladb-promoter commented:
CI state **`FAILURE`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/1634/

denesb commented:
@bhalevy 

```
[2022-08-02T08:35:45.268Z] failing test(s):
[2022-08-02T08:35:45.268Z] The following test(s) have failed: boost/sstable_datafile_test
```

bhalevy commented:
> @bhalevy
> 
> ```
> [2022-08-02T08:35:45.268Z] failing test(s):
> [2022-08-02T08:35:45.268Z] The following test(s) have failed: boost/sstable_datafile_test
> ```

@denesb I reproduced this on master, so it is preexisting.
I opened https://github.com/scylladb/scylladb/issues/11190 for it.

denesb commented:
Let's kick a new CI round then.

bhalevy commented:
> Let's kick a new CI round then.

https://jenkins.scylladb.com/job/releng/job/Scylla-CI/1643/

scylladb-promoter commented:
CI state **`SUCCESS`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/1643/
